### DISCUSSION TITLE
value converter with bindable ConverterParameter

### DISCUSSION THREAD

**Discussion by ali50m** [dotnet/wpf#8251] [Ideas]
It would be much more convenient if WPF can make the ConverterParameter bindable for value converter or introduce new contract for the converters. The value converter will be much more powerful then.

Not sure if it has been asked as I haven`t seen any results with search.

**Comment by lindexi** [comment]
@ali50m How about MultiBinding?

**Comment by ali50m** [comment]
@lindexi 不是很好用, 在执行Convert的时候有object[] values可以使用多个参数, 但是在执行IMultiValueConverter的ConvertBack的时候, 传入参数只有value (另外一个参数Type[] targetTypes几乎没什么用处), 容易缺少上下文信息. 如果有一个可供绑定的parameter,会直观和方便很多.

我遇到的实际问题如下, 我使用了一组RadioButton来表示页面导航按钮, RadioButton的IsChecked通过多值绑定到SelectedEnum和菜单Enum上. 在进行`ConvertBack`的时候, 希望可以将IsChecked的变化传回SelectedEnum. 但是能拿到的有意义的绑定值太少了.  试过用field或词典保存额外数据,但效果都不是很理想. 

如果有`ConverterParameter`这个绑定值, 那`ConvertBack`的可操作余地就会方便很多.

converter简化代码如下
```csharp
public sealed class NavigationMenuSelectionMultiConverter : IMultiValueConverter
{
    public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
    {
        if (values[0] is not { } selectedMenu)
            return false;

        if (values[1] is not { } navigationValue)
            return false;

        var b = selectedMenu.Equals(navigationValue);
        return b;
    }

    public object[] ConvertBack(
        object value,
        Type[] targetTypes,
        object parameter,
        CultureInfo culture
    )
    {
        return value is true
            // 这里不好实现, 缺乏上下文绑定信息
            ? new[] { _navigationValue ?? Binding.DoNothing, Binding.DoNothing }
            : new[] { Binding.DoNothing, Binding.DoNothing };
    }
}
```

内容比较多, 怕用英文说不太清除. 如果有需要, 我能提供一部分的英文信息.

**There is a lot of content, so I'm afraid it won't be clear enough with bad English. If necessary, I can try to provide part of the information in English.**

**Reply by lindexi** [reply]
@ali50m 

虽然有理由，但是我认为这不是一个很通用的使用情况哈  
There are reasons, but I don't think this is a very general use case.

也许考虑在 Converter 里面定义属性，通过属性传递参数。但是这样的缺点是要求在资源字典定义它的时候就对属性进行赋值，约等于不是一个绑定变量，也丢失绑定变量的功能  
Perhaps consider defining attributes in the Converter to pass parameters through. However, this has the disadvantage of requiring the attribute to be assigned when the resource dictionary defines it, which is equivalent to not being a bound variable and losing the function of binding variables

**Reply by ali50m** [reply]
@lindexi 我只是举了一个自身遇到的实际问题, 碰到了converter在使用时的局限性. 我相信应该会有不少人在其它场景同样被这个功能限制了, 或是使用其它不那么美好的方法绕过去了.

如果wpf的一大特色是数据绑定的话, 我觉得converter的parameter真的应该被放入到binding设计考虑中.

**Reply by lindexi** [reply]
@ali50m 

从复杂度的角度上考虑，我比较不倾向于在这里再次引入绑定。因为这样可以让一些有趣的开发者写出来像火车一样长的绑定代码，也就是在绑定里面的参数嵌套绑定。相信我，这个功能开放出来，肯定会有写出这样的代码的时候  
I am concerned that nested binding code will increase code complexity

**Reply by ali50m** [reply]
@lindexi 是否可以在不修改原有converter的情况下, 加入另外一套带有bindable parameter的converter(单值/多值). 
这样不关心这个变更的用户在使用原有converter的时候并不会获得 bindable parameter的功能. 知道这个变化的开发者, 大概率会是相对资深的wpf开发者(2023年还在关注wpf更新的开发者). 

另外,能使用bindable parameter, 对自定义控件的模板也是很有利的影响. 如果在控件模板或者style里提前实现了多个值的绑定, 也能在一定程度上影响用户在最终界面上使用冗长binding的几率.

**Reply by lindexi** [reply]
@ali50m 我感觉大概率这个 API 提议不会通过的哈  
I'm afraid the API proposal won't pass

**Reply by ali50m** [reply]
@lindexi 只是个建议, wpf这碗冷饭其实真的还能再炒一炒.

**Reply by lindexi** [reply]
@ali50m 哈哈，你看，我还是给你赞的投票

**Comment by ali50m** [comment]
另外,如果在Style或者controlTemplate里使用converter, 而parameter不能被绑定的话, 那converter就失去了一部分传入绑定信息的功能了.
